### ISSUE TITLE
Slicing with zero-length-slice fails

### ISSUE THREAD

**Issue by mrocklin** [pydata/sparse#276]
This happens a bunch in Dask Array because we like passing around empty versions of array chunks as metadata.

```python
import numpy as np, sparse
s = sparse.COO.from_numpy(np.ones((1)))
s[0:0]
```

### Traceback

<details>

```python-traceback
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-3-a774bf7a389e> in <module>
----> 1 s[0:0]

~/miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py in getitem(x, index)
     66
     67     # Get the mask
---> 68     mask, adv_idx = _mask(x.coords, index, x.shape)
     69
     70     # Get the length of the mask

~/miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py in _mask(coords, indices, shape)
    140         return mask, _AdvIdxInfo(aidxs, adv_idx_pos, len(adv_idx))
    141
--> 142     mask, is_slice = _compute_mask(coords, _ind_ar_from_indices(indices))
    143
    144     if is_slice:

~/miniconda/envs/dev/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    346                 e.patch_message(msg)
    347
--> 348             error_rewrite(e, 'typing')
    349         except errors.UnsupportedError as e:
    350             # Something unsupported is present in the user code, add help info

~/miniconda/envs/dev/lib/python3.7/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    313                 raise e
    314             else:
--> 315                 reraise(type(e), e, None)
    316
    317         argtypes = []

~/miniconda/envs/dev/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    656             value = tp()
    657         if value.__traceback__ is not tb:
--> 658             raise value.with_traceback(tb)
    659         raise value
    660

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function searchsorted at 0x10f763b70>) with argument(s) of type(s): (array(int64, 1d, C), int64, side=Literal[str](left))
 * parameterized
In definition 0:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function _searchsorted.<locals>.searchsorted_inner at 0x6148d66a8>) with argument(s) of type(s): (array(int64, 1d, C), int64)
 * parameterized
In definition 0:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<ufunc 'isnan'>) with argument(s) of type(s): (int64)
 * parameterized
In definition 0:
    TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typing/npydecl.py:110
In definition 1:
    TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typing/npydecl.py:110
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<ufunc 'isnan'>)
[2] During: typing of call at /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py (1997)


File "../../miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py", line 1997:
    def searchsorted_inner(a, v):
        <source elided>
        n = len(a)
        if np.isnan(v):
        ^

    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typeinfer.py:860
In definition 1:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<ufunc 'isnan'>) with argument(s) of type(s): (int64)
 * parameterized
In definition 0:
    TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typing/npydecl.py:110
In definition 1:
    TypingError: ufunc 'isnan' using the loop 'l->?' not supported in this mode
    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typing/npydecl.py:110
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<ufunc 'isnan'>)
[2] During: typing of call at /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py (1997)


File "../../miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py", line 1997:
    def searchsorted_inner(a, v):
        <source elided>
        n = len(a)
        if np.isnan(v):
        ^

    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typeinfer.py:860
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function _searchsorted.<locals>.searchsorted_inner at 0x6148d66a8>)
[2] During: typing of call at /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py (2052)


File "../../miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py", line 2052:
        def searchsorted_impl(a, v, side='left'):
            return loop_impl(a, v)
            ^

    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/typeinfer.py:860
In definition 1:
    ValueError: Invalid value given for 'side': unicode_type
    raised from /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/numba/targets/arraymath.py:2030
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function searchsorted at 0x10f763b70>)
[2] During: typing of call at /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py (451)


File "../../miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py", line 451:
def _get_mask_pairs(starts_old, stops_old, c, idx):  # pragma: no cover
    <source elided>
        for p_match in range(idx[0], idx[1], idx[2]):
            start = np.searchsorted(c[starts_old[j]:stops_old[j]], p_match, side='left') + starts_old[j]
            ^

[1] During: resolving callee type: type(CPUDispatcher(<function _get_mask_pairs at 0x114264ea0>))
[2] During: typing of call at /Users/mrocklin/miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py (387)


File "../../miniconda/envs/dev/lib/python3.7/site-packages/sparse/coo/indexing.py", line 387:
def _compute_mask(coords, indices):  # pragma: no cover
    <source elided>
        # Which would come out of indexing a single integer.
        starts, stops, n_matches = _get_mask_pairs(starts, stops, coords[i], indices[i])
        ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

</details>

**Comment by hameerabbasi** [comment]
In theory, this should be solved by #264, but Numba’s typed lists are a WIP and have bugs.

**Comment by hameerabbasi** [comment]
@mrocklin How urgent is this? I could just add a special-case for now.

**Comment by hameerabbasi** [comment]
Solved with a test added in #264.
